pandas/core/groupby/groupby.py

```diff
@@ -1023,7 +1025,14 @@ def _set_group_selection(self) -> None:
         if len(groupers):
             # GH12839 clear selected obj cache when group selection changes
             ax = self.obj._info_axis
-            self._group_selection = ax.difference(Index(groupers), sort=False).tolist()
+            if len(ax) < 1000:
```

Comment:
Generally I don't like constant thresholds like this as it's not guaranteed to hold overtime and is susceptible to behavior difference that are data size related (which can be hard to debug).

I would prefer just to use one path.

Reply:
Thanks - I found a more efficient implementation, timings are in https://github.com/pandas-dev/pandas/pull/50846#discussion_r1087198598

Reply:
Awesome!